tests/test_schedules.py

```diff
@@ -1132,6 +1132,188 @@ def foo(a: f32 @ DRAM, b: f32 @ DRAM):
         foo = reuse_buffer(foo, "bb:_", "c:_")
 
 
+def test_fold_buffer_loop_simple(golden):
+    @proc
+    def foo(N: size):
+        assert N > 4
+        x: i8[N]
+        for i in seq(0, N - 4):
+            for j in seq(i, i + 4):
+                x[j] = 1.0
+
+    with pytest.raises(SchedulingError, match="Buffer folding not possible"):
+        foo = fold_buffer(foo, foo.find("x: _"), 0, 2)
+
+    foo = fold_buffer(foo, foo.find("x: _"), 0, 3)
+    foo = simplify(foo)
+    assert str(foo) == golden
+
+
+# TODO: to make this kind of test work, we need to support more general
+# index analysis which potentially leverages SMT solvers for comparisons.
+# def test_fold_buffer_loop_in_context(golden):
+#     @proc
+#     def foo(N: size):
+#         assert N > 2
+#         x: i8[N]
+#         x[2] = 0.0
+#         for i in seq(0, N / 2):
+#             x[2 * i] = 1.0
+#             x[2 * i + 1] = 2.0
+#         x[N / 2] = 3.0
+
+#     with pytest.raises(SchedulingError, match="Buffer folding not possible"):
+#         foo = fold_buffer(foo, foo.find("x: _"), 0, 2)
+
+#     foo = fold_buffer(foo, foo.find("x: _"), 0, 3)
+#     foo = simplify(foo)
+#     assert str(foo) == golden
+
+
+def test_fold_buffer_sequential_stmts(golden):
+    @proc
+    def foo():
+        x: i8[10]
+        x[0] = 0.0
+        x[2] = 0.0
+        x[3] = 0.0
+        x[2] = 0.0
+        x[7] = 0.0
+        x[5] = 0.0
```

Comment:
This should succeed, right?
```
    @proc
    def foo():
        x: i8[20]
        x[0] = 0.0
        x[2] = 0.0
        x[3] = 0.0
        x[2] = 0.0
        x[7] = 0.0
        x[5] = 0.0
```
`foo = fold_buffer(foo, foo.find("x: _"), 0, 10)`